### PR TITLE
Generate docs with sphinx

### LINKED ISSUE #48: Use sphinx to enable cross referencing
In the meeting it was suggested to consider using sphinx (as described in the existing documentation design document). That would allow to cross reference specific parts of the generated documentation instead of relying on "full" URLs.

### PR DESCRIPTION
**Pull request by apojomovsky**
Fixes #48 
Makes use of Sphinx to generate embeddable HTML pages from markdown files , relegating Jekyll from the conversion responsibility.
This enables us to cross-reference specific parts of the generated documentation.

### CODE REVIEW

File: _plugins/docs_generator.rb [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    end
+  end
+
+  def search_docfiles(doc_repos_hash)
+    docfiles_hash = Hash.new
+    doc_repos_hash.each do |repo_name, repo|
+      docs_in_repo = Array.new
+      Dir.glob(File.join("_remotes", repo_name) + '**/*.{md, rst}', File::FNM_CASEFOLD) do |doc_relpath|
+        docs_in_repo.push(doc_relpath)
+      end
```
Comment (hidmic) [inline _plugins/docs_generator.rb]: @apojomovsky nit: I believe you can just do `docs_in_repo = Dir.glob(...)`

File: _plugins/docs_generator.rb [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+      docfiles_hash[repo_name] = docs_in_repo
+    end
+    docfiles_hash
+  end
+
+  def convert_with_sphinx(docfiles)
+    json_files_data = Hash.recursive
+    docfiles.each do |repo_name, files|
+      command = "sphinx-build -b json -c _sphinx #{File.join('_sphinx', 'repos', repo_name)} _sphinx/_build"
+      puts command
```
Comment (hidmic) [inline _plugins/docs_generator.rb]: @apojomovsky nit: extra `puts`?

File: _plugins/docs_generator.rb [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      end
+      docfiles_hash[repo_name] = docs_in_repo
+    end
+    docfiles_hash
+  end
+
+  def convert_with_sphinx(docfiles)
+    json_files_data = Hash.recursive
+    docfiles.each do |repo_name, files|
+      command = "sphinx-build -b json -c _sphinx #{File.join('_sphinx', 'repos', repo_name)} _sphinx/_build"
```
Comment (hidmic) [inline _plugins/docs_generator.rb]: @apojomovsky nit: consider defining a `repo_path = File.join(...)` for better readability.

File: _plugins/docs_generator.rb [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      puts command
+      pid = Kernel.spawn(command)
+      Process.wait pid
+      files.each do |file|
+        name = File.basename(file, File.extname(file))
+        if name.scan(/readme/i).length > 0
+          name = "index"
+        end
+        json_file = File.join('_sphinx/_build/', name + '.fjson')
+        next unless File.file?(json_file)
```
Comment (hidmic) [inline _plugins/docs_generator.rb:75]: @apojomovsky can't we just iterate over a glob of JSON files?

File: _plugins/docs_generator.rb [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        json_content = File.read(json_file)
+        json_files_data[repo_name][name] = JSON.parse(json_content)
+
+        # Generate HTML tables that were left with markdown syntax by sphinx
```
Comment (hidmic) [inline _plugins/docs_generator.rb]: @apojomovsky nit: consider adding a reference to the missing feature tracking issue.

File: _plugins/docs_generator.rb [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        html_doc = Nokogiri::HTML(JSON.parse(json_content)["body"])
+        html_doc.css('p').each do |paragraph|
+          if paragraph.to_s[3..-1].strip[0] == "|"
```
Comment (hidmic) [inline _plugins/docs_generator.rb]: @apojomovsky I think checking `paragraph.content.strip[0] == "|"` would be more robust.

File: _plugins/docs_generator.rb [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            html_table = Kramdown::Document.new(paragraph.to_s.gsub("<p>", "").gsub("</p>", ""))
```
Comment (hidmic) [inline _plugins/docs_generator.rb]: @apojomovsky same as above, checking `paragraph.content` is more robust (as you may be dealing with a `<p>` tag with attributes).

File: _plugins/docs_generator.rb [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+            paragraph.replace(html_table.to_html.gsub("<table>", "<table border=\"1\" class=\"docutils>\""))
```
Comment (hidmic) [inline _plugins/docs_generator.rb]: @apojomovsky consider verifying that what you have is actually a table. Maybe using `nokogiri` again? You may also add any extra attributes this way.

File: _plugins/docs_generator.rb [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+          end
+        end
+        json_files_data[repo_name][name]["body"] = html_doc.to_s.lines[2..-2].join
```
Comment (hidmic) [inline _plugins/docs_generator.rb]: @apojomovsky why are we dropping some lines?

File: _plugins/editable_docs.rb [hunk 10]
```diff
@@ -7,13 +7,18 @@
...
     next if repo_name.nil? or repo_data.nil?
 
     # Generate edit url
-    page_relative_url = page_path.sub(File.join("doc", repo_name), "")
-    page.data["edit_url"] = generate_edit_url(repo_data, page_relative_url)
+    edit_relative_url = page_path.sub(File.join("doc", repo_name), "")
+    if edit_relative_url.scan(/\//).length > 1
+      edit_relative_url = edit_relative_url[0..-12] + page.data["file_extension"]
+    else
+      edit_relative_url = "/README.md"
```
Comment (hidmic) [inline _plugins/editable_docs.rb]: @apojomovsky this isn't entirely true, it's equally likely to find an `index.rst` file. Why not keep the entire relative path as page data?

File: _plugins/editable_docs.rb [hunk 11]
```diff
@@ -7,13 +7,18 @@
...
+      page_path.start_with?(File.join("doc", name))
     end
     next if repo_name.nil? or repo_data.nil?
 
     # Generate edit url
-    page_relative_url = page_path.sub(File.join("doc", repo_name), "")
-    page.data["edit_url"] = generate_edit_url(repo_data, page_relative_url)
+    edit_relative_url = page_path.sub(File.join("doc", repo_name), "")
+    if edit_relative_url.scan(/\//).length > 1
+      edit_relative_url = edit_relative_url[0..-12] + page.data["file_extension"]
```
Comment (hidmic) [inline _plugins/editable_docs.rb]: @apojomovsky I don't entirely get why the slicing.

File: _sphinx/conf.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+# for a list of supported languages.
+#
+# This is also used if you do content translation via gettext catalogs.
+# Usually you set "language" from the command line for these cases.
+language = None
+
+# List of patterns, relative to source directory, that match files and
+# directories to ignore when looking for source files.
+# This patterns also effect to html_static_path and html_extra_path
+exclude_patterns = ['_build', '_scripts', '_cache', '_devel', '_plugins_data', '_ruby_libs', 'node_modules', 'js', 'Thumbs.db', '.DS_Store']
```
Comment (hidmic) [inline _sphinx/conf.py]: @apojomovsky I don't get the rationale of this setting. Isn't sphinx running on a subdirectory of Jekyll's source dir?

File: _sphinx/conf.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+# -- Options for Texinfo output -------------------------------------------
+
+# Grouping the document tree into Texinfo files. List of tuples
+# (source start file, target name, title, author,
+#  dir menu entry, description, category)
+texinfo_documents = [
+    (master_doc, 'ros2_docs', u'ros2_docs Documentation',
+     author, 'ros2_docs', 'One line description of project.',
+     'Miscellaneous'),
+]
```
Comment (hidmic) [inline _sphinx/conf.py]: @apojomovsky consider removing unused configuration settings (e.g. for tex output, for man pages output, etc.).

File: docker/Dockerfile [hunk 14]
```diff
@@ -34,6 +35,11 @@ RUN apt-get update && \
         tmux && \
     rm -rf /var/lib/apt/lists/*
 
+RUN pip install --upgrade setuptools pip
+RUN pip install \
+        recommonmark==0.4 \
+        sphinx==1.5.6
```
Comment (hidmic) [inline docker/Dockerfile:41]: @apojomovsky any powerful reason to get hooked to these specific versions?

File: _plugins/docs_generator.rb [hunk 15]
```diff
@@ -0,0 +1,109 @@
+require 'fileutils'
+require 'nokogiri'
+
+class DocPage < Jekyll::Page
```
Comment (hidmic) [inline _plugins/docs_generator.rb]: @apojomovsky consider moving this class to the `_ruby_libs/pages.rb` source file.

**Review by hidmic — CHANGES_REQUESTED** [review]
@apojomovsky I've left some comments. I think that, with this set of improvements, we can safely deprecate the `RstAsMarkdown` generator plugin. Awesome work!

File: _plugins/docs_generator.rb (hunk 1 above)
Comment (apojomovsky) [reply _plugins/docs_generator.rb]: that's a much cleaner approach. done, thanks!

File: _plugins/docs_generator.rb (hunk 2 above)
Comment (apojomovsky) [reply _plugins/docs_generator.rb]: oops, just removed it. thanks!

File: _plugins/docs_generator.rb (hunk 3 above)
Comment (apojomovsky) [reply _plugins/docs_generator.rb]: totally agree, done! thanks!

File: _plugins/docs_generator.rb (hunk 4 above)
Comment (apojomovsky) [reply _plugins/docs_generator.rb:75]: We could, but for the current approach, this is required to know the name of the original markdown file in order to save it as a page's data (for the Edit me on Github link)

File: _plugins/docs_generator.rb (hunk 5 above)
Comment (apojomovsky) [reply _plugins/docs_generator.rb]: done, thanks!

File: _plugins/docs_generator.rb (hunk 6 above)
Comment (apojomovsky) [reply _plugins/docs_generator.rb]: you're right, just did that. thanks!

File: _plugins/docs_generator.rb (hunk 7 above)
Comment (apojomovsky) [reply _plugins/docs_generator.rb]: The thing with `paragraph.content` is that it removes every `HTML` syntax from inside the block, making it unsuitable for our needs. I couldn't find another way to properly access it besides this one.

File: _plugins/docs_generator.rb (hunk 8 above)
Comment (apojomovsky) [reply _plugins/docs_generator.rb]: I added some custom logic to check for this now, thanks!

File: _plugins/docs_generator.rb (hunk 9 above)
Comment (apojomovsky) [reply _plugins/docs_generator.rb]: Yeah, funny thing here. The `HTML` generator spits a "complete" webpage with `<html>` and `<body>` tags that might break our destination file, so this is the way I found to remove them.

File: _plugins/editable_docs.rb (hunk 10 above)
Comment (apojomovsky) [reply _plugins/editable_docs.rb]: You're right. Just fixed the logic to keep track of the original file name and put it here. Thanks!

File: _plugins/editable_docs.rb (hunk 11 above)
Comment (apojomovsky) [reply _plugins/editable_docs.rb]: It was removing the `/index.html` part out of the string, but this code block is now gone.

File: _sphinx/conf.py (hunk 12 above)
Comment (apojomovsky) [reply _sphinx/conf.py]: You're right, just fixed this. Thanks!

File: docker/Dockerfile (hunk 14 above)
Comment (apojomovsky) [reply docker/Dockerfile:41]: Yes, there's some incompatibility across `recommonmark` and some newer versions of `sphinx`, so what I did here is to freeze `recommonmark` to its latest version to the date, and freeze `sphinx` to the latest known to work version.

File: _plugins/docs_generator.rb (hunk 15 above)
Comment (apojomovsky) [reply _plugins/docs_generator.rb]: done, thanks!

**Comment by apojomovsky** [comment]
Thanks a lot for the thorough review @hidmic . I just finished addressing your comments. Could you PTAL?

File: docker/Dockerfile (hunk 14 above)
Comment (hidmic) [reply docker/Dockerfile:41]: Makes sense.

File: _plugins/docs_generator.rb [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+
+        # Generate HTML tables that were left with markdown syntax by sphinx.
+        # Please refer to this issue for more information:
+        # https://github.com/rtfd/recommonmark/issues/3
+        # Although there's a sphinx-markdown-tables extension now, its
+        # functionality is limited and doesn't fit our requirements.
+        html_doc = Nokogiri::HTML(JSON.parse(json_content)["body"])
+        html_doc.css('p').each do |paragraph|
+          if paragraph.content.strip[0] == "|"
+            html_table = Kramdown::Document.new(paragraph.to_s[paragraph.to_s.index(/\|/)..-1].sub("</p>", "")).to_html
```
Comment (hidmic) [inline _plugins/docs_generator.rb]: @apojomovsky hmm yes, I see. Have you tried with `.inner_html`?

File: _plugins/docs_generator.rb [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+            if html_table.scan(/\<table/).length > 0
+              paragraph.replace(html_table.gsub("<table", "<table border=\"1\" class=\"docutils\" "))
```
Comment (hidmic) [inline _plugins/docs_generator.rb]: @apojomovsky I still think that using Nokogiri for this tinkering is way more robust.

File: _plugins/docs_generator.rb [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+              json_files_data[repo_name][name]["body"] = html_doc.to_s.lines[2..-2].join
```
Comment (hidmic) [inline _plugins/docs_generator.rb]: @apojomovsky Oh, hmm, Nokogiri to remove :) ?

Comment (hidmic) [reply _plugins/docs_generator.rb]: BTW shouldn't this statement be outside the loop?

File: _plugins/editable_docs.rb [hunk 19]
```diff
@@ -7,17 +7,16 @@
...
     repo_name, repo_data = repositories.find do |name, data|
-      page_path.start_with? (File.join("doc", name))
+      page_path.start_with?(File.join("doc", name))
     end
     next if repo_name.nil? or repo_data.nil?
-
     # Generate edit url
-    page_relative_url = page_path.sub(File.join("doc", repo_name), "")
-    page.data["edit_url"] = generate_edit_url(repo_data, page_relative_url)
+    original_filepath = page.data["file_relpath"].split('/')[2..-1].join('/')
```
Comment (hidmic) [inline _plugins/editable_docs.rb]: @apojomovsky why the new slicing here?

**Review by hidmic — CHANGES_REQUESTED** [review]
@apojomovsky one last pass!

File: _plugins/docs_generator.rb (hunk 16 above)
Comment (apojomovsky) [reply _plugins/docs_generator.rb]: that worked just great, thanks!

File: _plugins/docs_generator.rb (hunk 17 above)
Comment (apojomovsky) [reply _plugins/docs_generator.rb]: just did as suggested, works great! thanks.

File: _plugins/docs_generator.rb (hunk 18 above)
Comment (apojomovsky) [reply _plugins/docs_generator.rb]: oops, true. just put it outside. thanks

File: _plugins/editable_docs.rb (hunk 19 above)
Comment (apojomovsky) [reply _plugins/editable_docs.rb]: This takes a path that is relative to the project's root and "goes up" two levels 
`_remotes/ros2_overview/Build-Cop-and-Build-Farmer-Guide.md` --> `Build-Cop-and-Build-Farmer-Guide.md`.
But I just revisited this code to make it more robust by using a sub-string search.

**Comment by apojomovsky** [comment]
Thanks for the review @hidmic! Comments addressed, PTAL

**Review by hidmic — APPROVED** [review]

**Pull request merged**